### PR TITLE
Allow users to specify the processed files via a regex

### PR DESCRIPTION
**Pull request by JBludau**
This PR adds a user option (via environment vars) to specify which files should be processed. For this it uses the regex option of find.

This came up since we also use `.cmake.in` files and wanted them to be formatted as well

### CODE REVIEW

**Comment by puneetmatharu** [comment]
Hi @JBludau, thanks for your contribution! It's certainly a useful feature. Could you update the `README.md` to show users how to use this new feature? Thanks.

**Comment by JBludau** [comment]
Hi @puneetmatharu, I reworked it a bit to be able to supply both the regex and the arguments to `cmake-format` as command-line arguments. Since`sh` is really hard to keep from wanting to interpret the passed regex as individual arguments and since we have `python` available I opted for using it directly. It might be overkill but it allowed to separate the two groups of arguments.

Is there a way to test if the usage example I put in the README is working with actions? I was only able to test if it works if I pass the arguments to docker run e.g.
```
docker run -e GITHUB_WORKSPACE=/mount -v .:/mount <ImageID> --file_regex='(.*\.cmake\.in$|.*\.cmake$|CMakeLists.txt$)' --cmake_format_args='--testarg'
```

File: entrypoint.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/python3
+
+import os
+import argparse
+import re
+import subprocess
```
Comment (romintomasetti) [inline entrypoint.py]: Order the includes :wink:

File: entrypoint.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+                    prog='ProgramName',
+                    description='What the program does',
+                    epilog='Text at the bottom of help')
```
Comment (romintomasetti) [inline entrypoint.py]:
```suggestion
    parser = argparse.ArgumentParser()
```

File: entrypoint.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    parser.add_argument('--file_regex')
```
Comment (romintomasetti) [inline entrypoint.py]:
```suggestion
    parser.add_argument('--file-regex')
```
and then you can access the variable with `args.file_regex`.

Comment (JBludau) [reply entrypoint.py]: Didn't know that argparse replaces `-` with `_` in the variable name

Comment (JBludau) [reply entrypoint.py]: did it for all occurrences

File: entrypoint.py (hunk 1 above)
Comment (JBludau) [reply entrypoint.py]: done in https://github.com/puneetmatharu/cmake-format-lint-action/pull/9/commits/eb98f0c07e7ec986e762bddfcbda39665ef3398b

File: entrypoint.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#!/bin/python3
+
```
Comment (puneetmatharu) [inline entrypoint.py:2]: I really like the idea to use Python3. If we're going with it, how about we do it like the following:

```py
#!/usr/bin/env python3

import argparse
import os
import re
import subprocess
from pathlib import Path
from multiprocessing import Pool
from typing import List, Tuple, Optional


def format_cmake_file(file_and_args: Tuple[Path, str]) -> Optional[str]:
    (file_path, cmake_format_args) = file_and_args
    cmake_command = f"cmake-format {cmake_format_args} {file_path}"
    cmake_format_result = subprocess.run(cmake_command, shell=True, capture_output=True, text=True)
    if cmake_format_result.returncode != 0:
        return f"Error formatting {file_path}: {cmake_format_result.stderr}"
    return None


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Format CMake files using cmake-format.")
    parser.add_argument('--file-regex', required=True, help="Regex pattern to match filenames.")
    parser.add_argument('--cmake-format-args', required=True, help="Additional arguments for cmake-format.")
    args = parser.parse_args()
    if len(args.file_regex) == 0:
        parser.error("The --file-regex argument cannot be empty.")
    return args


if __name__ == "__main__":
    args = parse_args()

    file_regex = re.compile(args.file_regex)
    workspace_path = Path(os.environ['GITHUB_WORKSPACE'])

    matched_files = [
        (file_path, args.cmake_format_args)
        for file_path in workspace_path.rglob('*')
        if file_path.is_file() and file_regex.match(file_path.name)
    ]

    if len(matched_files) > 0:
        with Pool() as pool:
            format_results = pool.map(format_cmake_file, matched_files)
        for result in format_results:
            if result:
                print(result)
    else:
        print("No files matched the given regex.")
```

to be clearer, catch empty regex strings, and to process files in parallel.

File: README.md [hunk 5]
```diff
@@ -6,7 +6,8 @@ Formats CMake-specific files to the desired format using [`cmake-format`](https:
 
 ## Usage
 
-To use this action, pass arguments to the `args` element as you would to `cmake-format` - these arguments will be used to format each CMake file. For example
+To use this action, pass arguments to the `args` element as you would to `cmake-format` - these arguments will be used to format each CMake file. To select the files which should be processed use the `find_regex` argument.
```
Comment (puneetmatharu) [inline README.md]: How about we change 
```md
To select the files which should be processed use the `find_regex` argument.
```
to
```
For more fine-grained control over which files to process, use the `find_regex` argument; see below for an example.
```

**Review by puneetmatharu — CHANGES_REQUESTED** [review]
This is great, and I really like the idea to move to Python. Nice thinking. Please see my only two remarks for changes. After, I would be happy to merge things in.

File: README.md (hunk 5 above)
Comment (JBludau) [reply README.md]: added in https://github.com/puneetmatharu/cmake-format-lint-action/pull/9/commits/c4c94b0b769e7fdc2430fcb79fdf2ff5e2d663b6

File: entrypoint.py (hunk 4 above)
Comment (JBludau) [reply entrypoint.py:2]: sure, done in  https://github.com/puneetmatharu/cmake-format-lint-action/pull/9/commits/a64400ece7172ecdae85fbfc9006348f32a71ef2

**Comment by JBludau** [comment]
should we bump the version number?

**Comment by puneetmatharu** [comment]
Hi @JBludau, I'm going to close this PR. I've patched the `action.yaml` file and swapped to using a composite action instead of a Docker-based one so it can also be used on macOS. You should be able to test out these features with the v1.0.5 tag. Let me know if you have any issues. Thanks!

**Comment by JBludau** [comment]
Yeah, it absolutely makes sense to get rid of the docker redirection. The v1.0.5 works for us. Thanks for the support!

**Pull request closed without merging**